File: app_mobile/gest_squadre/lib/services/map_view_storage.dart
import 'package:shared_preferences/shared_preferences.dart';

import 'map_layer_config.dart';

class MapViewState {
  const MapViewState({
    required this.latitude,
    required this.longitude,
    required this.zoom,
    required this.layerMode,
  });

  final double latitude;
  final double longitude;
  final double zoom;
  final MapLayerMode layerMode;
}

class MapViewStorage {
  static const _latKey = 'gest_toc_map_lat';
  static const _lngKey = 'gest_toc_map_lng';
  static const _zoomKey = 'gest_toc_map_zoom';
  static const _layerKey = 'gest_toc_map_layer';

  static const defaultLat = 45.0703;
  static const defaultLng = 7.6869;
  static const defaultZoom = 13.0;

  Future<MapViewState> load() async {
    final prefs = await SharedPreferences.getInstance();
    return MapViewState(
      latitude: prefs.getDouble(_latKey) ?? defaultLat,
      longitude: prefs.getDouble(_lngKey) ?? defaultLng,
      zoom: prefs.getDouble(_zoomKey) ?? defaultZoom,
      layerMode: MapLayerConfig.fromStorage(prefs.getString(_layerKey)),
    );
  }

  Future<void> saveView({
    required double latitude,
    required double longitude,
    required double zoom,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setDouble(_latKey, latitude);
    await prefs.setDouble(_lngKey, longitude);
    await prefs.setDouble(_zoomKey, zoom);
  }

  Future<void> saveLayer(MapLayerMode mode) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_layerKey, MapLayerConfig.toStorage(mode));
  }
}

File: app_mobile/kmp/androidApp/src/main/java/com/ansmi/gestsquadre/kmp/map/MapViewStorage.kt
package com.ansmi.gestsquadre.kmp.map

import android.content.Context

data class MapViewState(
    val latitude: Double,
    val longitude: Double,
    val zoom: Double,
    val layerMode: MapLayerMode,
) {
    companion object {
        const val DEFAULT_LAT = 45.0703
        const val DEFAULT_LNG = 7.6869
        const val DEFAULT_ZOOM = 13.0

        val DEFAULT =
            MapViewState(
                latitude = DEFAULT_LAT,
                longitude = DEFAULT_LNG,
                zoom = DEFAULT_ZOOM,
                layerMode = MapLayerMode.STANDARD,
            )
    }
}

enum class MapLayerMode {
    STANDARD,
    ORTHOPHOTO,
    ;

    companion object {
        fun fromStorage(raw: String?): MapLayerMode =
            if (raw == "orthophoto") ORTHOPHOTO else STANDARD

        fun toStorage(mode: MapLayerMode): String =
            if (mode == ORTHOPHOTO) "orthophoto" else "standard"
    }
}

class MapViewStorage(
    context: Context,
) {
    private val prefs = context.getSharedPreferences("gest_toc_map", Context.MODE_PRIVATE)

    fun load(): MapViewState =
        MapViewState(
            latitude = prefs.getFloat(KEY_LAT, MapViewState.DEFAULT_LAT.toFloat()).toDouble(),
            longitude = prefs.getFloat(KEY_LNG, MapViewState.DEFAULT_LNG.toFloat()).toDouble(),
            zoom = prefs.getFloat(KEY_ZOOM, MapViewState.DEFAULT_ZOOM.toFloat()).toDouble(),
            layerMode = MapLayerMode.fromStorage(prefs.getString(KEY_LAYER, null)),
        )

    fun saveView(
        latitude: Double,
        longitude: Double,
        zoom: Double,
    ) {
        prefs.edit()
            .putFloat(KEY_LAT, latitude.toFloat())
            .putFloat(KEY_LNG, longitude.toFloat())
            .putFloat(KEY_ZOOM, zoom.toFloat())
            .apply()
    }

    fun saveLayer(mode: MapLayerMode) {
        prefs.edit()
            .putString(KEY_LAYER, MapLayerMode.toStorage(mode))
            .apply()
    }

    companion object {
        private const val KEY_LAT = "gest_toc_map_lat"
        private const val KEY_LNG = "gest_toc_map_lng"
        private const val KEY_ZOOM = "gest_toc_map_zoom"
        private const val KEY_LAYER = "gest_toc_map_layer"
    }
}
